source/grid/tria.cc

```diff
@@ -15165,6 +15168,157 @@ Triangulation<1, 3>::prepare_coarsening_and_refinement()
+  {
+    // We will access the indices of vertices of cells below. For
+    // this to work, we have to first make sure the cache for this
+    // information is up to date.
+    reset_cell_vertex_indices_cache();
+
+    unsigned int next_unused_index = 0;
+
+    // First the active cell iterators:
+    for (const auto &cell : active_cell_iterators())
+      for (const unsigned int v : cell->vertex_indices())
+        {
+          const unsigned int vertex_index = cell->vertex_index(v);
+          if (old_to_new_vertex_indices[vertex_index] ==
+              numbers::invalid_unsigned_int)
+            old_to_new_vertex_indices[vertex_index] = next_unused_index++;
+        }
+    // Then the non-active cell iterators (we already covered the active
+    // ones):
+    for (const auto &cell : cell_iterators())
+      if (cell->is_active() == false)
+        for (const unsigned int v : cell->vertex_indices())
+          {
+            const unsigned int vertex_index = cell->vertex_index(v);
+            if (old_to_new_vertex_indices[vertex_index] ==
+                numbers::invalid_unsigned_int)
+              old_to_new_vertex_indices[vertex_index] = next_unused_index++;
+          }
+
+    // At this point, we should have covered all used vertices. Check this:
+    for (unsigned int v = 0; v < vertices.size(); ++v)
+      Assert(((vertices_used[v] == true) && (old_to_new_vertex_indices[v] !=
+                                             numbers::invalid_unsigned_int)) ||
+               ((vertices_used[v] == false) && (old_to_new_vertex_indices[v] ==
+                                                numbers::invalid_unsigned_int)),
+             ExcInternalError());
+
+    // This leaves the unused vertices. Enumerate them as well:
+    for (unsigned int v = 0; v < vertices.size(); ++v)
+      if (old_to_new_vertex_indices[v] == numbers::invalid_unsigned_int)
+        old_to_new_vertex_indices[v] = next_unused_index++;
+    Assert(next_unused_index == vertices.size(), ExcInternalError());
+  }
+
+
+  // Step 2: Now we need to reorder or reindex all sorts of data structures.
+  // This can all be done on parallel tasks.
+  Threads::TaskGroup<> tasks;
+
+  // Reorder the vertices[] array:
+  tasks += Threads::new_task(
+    [&old_to_new_vertex_indices, &vertices = this->vertices]() {
+      std::vector<Point<spacedim>> new_vertices(vertices.size());
+      for (unsigned int v = 0; v < old_to_new_vertex_indices.size(); ++v)
+        new_vertices[old_to_new_vertex_indices[v]] = vertices[v];
+      vertices = std::move(new_vertices);
+    });
+
+  // Reorder the vertices_used[] array:
+  tasks += Threads::new_task(
+    [&old_to_new_vertex_indices, &vertices_used = this->vertices_used]() {
+      std::vector<bool> new_vertices_used(vertices_used.size());
+      for (unsigned int v = 0; v < old_to_new_vertex_indices.size(); ++v)
+        new_vertices_used[old_to_new_vertex_indices[v]] = vertices_used[v];
```

Comment:
The operations are not that independent, so let me give some background to what @tamiko wrote above regarding the expected speedup: We could do a little better by not running two separate loops for `new_vertices_used` and `new_vertices` in line 15245 above, but do both operations in the same loop, at essentially the same cost. This way, the loop overhead is only paid once, and `old_to_new_vertex_indices` is read only once rather than twice. If memory transfer is the bottleneck, which it should be at least for the `new_vertices` field, the sequential version can be quicker. The reality is that one would need to benchmark the two variants and spend time to see which version is actually beneficial, I do not think it is straight-forward with the present parallelization by tasks. I will open an issue with more details (but probably not today but tomorrow because I need to rush away).

Reply:
Fair point. I'll be happy to merge these two lambdas. 

In any case, as I said, there are two tasks for now: (i) get it to work correctly, (ii) agree or not whether this is the right way to go to begin with.

Reply:
The merged loop is in the next commit I'll push.